### ISSUE TITLE
AN can't boot anything, linux ext4, btrfs, macOS Monterey, Big Sur, Catalina

### ISSUE THREAD

**Issue by MarioG-X** [dakanji/RefindPlus#94]
### RefindPlus Version

v0.13.2.AN Release

### Device Type

Apple Mac

### Problem Description

Upgrade from AM RPF and conf. Nothing can't boot anymore. The attached logs are from an attempt to boot Monterey, then OPenSUSE BTRFS. Monterey: press key while Monterey selected and screen does not change. Opensuse: Press key and RFP loader message displayd but change there, no linux screen.

I also tried ext4 linux and got same results. Also tried Catalina and screen went white. No logs for these unless they still have an issue after this issue is fixed.

### Problem Point

While starting a loader or tool

### Affected Items

All OS Loaders (First Row Items)

### Debug Log

Ran with dbg and logelevel=q
[22d08y0734.log](https://github.com/dakanji/RefindPlus/files/8027890/22d08y0734.log)
[22d08y0836.log](https://github.com/dakanji/RefindPlus/files/8027892/22d08y0836.log)

### Additional Context

Either case never leaves RFP screen.

**Comment by dakanji** [comment]
Odd. Perhaps an issue with the uploaded file.
Please try X198

[X198-BOOTx64.zip](https://github.com/dakanji/RefindPlus/files/8029483/X198-BOOTx64.zip)

**Comment by dakanji** [comment]
OK. Thanks. Will look into this

**Comment by startergo** [comment]
This one boots, but now shell.efi does not boot.
[Archive.zip](https://github.com/dakanji/RefindPlus/files/8036819/Archive.zip)
.

**Comment by dakanji** [comment]
Yes. Shell will not boot as text render is missing.
I guess we can continue looking at this here. So issues arose between this and the other commit.

Will take some digging to unravel as the likely offending commit was a big lump.

**Comment by dakanji** [comment]
Please try X199b to confirm breaking commit
[X199b-BOOTx64.efi.zip](https://github.com/dakanji/RefindPlus/files/8036877/X199b-BOOTx64.efi.zip)

**Comment by startergo** [comment]
Not booting.
[22d09z4626.log.zip](https://github.com/dakanji/RefindPlus/files/8036905/22d09z4626.log.zip)

**Comment by dakanji** [comment]
Thanks. Break was in that commit.
Need to Isolate the the breaking file and I am getting the punishment for lumping too much into one commit.
Can then look for the breaking line(s) in that file

Will need a few iterations

**Comment by dakanji** [comment]
Ok. Please try X200a
[X200a-BOOTx64.efi.zip](https://github.com/dakanji/RefindPlus/files/8037068/X200a-BOOTx64.efi.zip)

**Comment by MarioG-X** [comment]
X200a same:
[22d10a2950.log](https://github.com/dakanji/RefindPlus/files/8037099/22d10a2950.log)

**Comment by dakanji** [comment]
Just to be sure, does this X119a work for you as well: https://github.com/dakanji/RefindPlus/issues/94#issuecomment-1034303426?

**Comment by MarioG-X** [comment]
X199a fails same place:
[22d10a3845.log](https://github.com/dakanji/RefindPlus/files/8037122/22d10a3845.log)

**Comment by dakanji** [comment]
Hmm. It apparently worked for @startergo

Did you get a firmware update with a MacOS update in the past couple of months by any chance?

**Comment by startergo** [comment]
He is booting Linux, which does not boot. I was never able to boot Linux without Grubx64.efi

**Comment by dakanji** [comment]
vmlinuz works ok and @MarioG-X has been using this all along.
Just now think the issues are memory related and could be tied to the recent firmware update issues.
I am not having any of these boot issues on my machine.

@MarioG-X ... maybe best to stick with trying MacOS first for the time being.
Still curious as I suppose you must have updated Monterey recently and most likely got the firmware update that has been causing problems.

**Comment by MarioG-X** [comment]
Don’t think any firmware updates affected this since AM still works fine.

btw, I don’t use grub anywhere. Not even installed on any Linux.

**Comment by dakanji** [comment]
> Don’t think any firmware updates affected this since AM still works fine

AM is not affected because it disabled virtually all memory freeing

**Comment by dakanji** [comment]
Let me do a build of AN that also disables this

**Comment by MarioG-X** [comment]
The only Monterey firmware issues I could find related to macs with T2 chips. This one is not a T2, it is a 2014 Mac mini. I have had no other issues with the machine. AM has zero issues on it.

**Comment by dakanji** [comment]
> AM has zero issues on it.

As said, AM is not affected for the reason stated as I specifically tweaked the AM release files for this reason.

There was a firmware update that has created various issues and it is not T2 related.
Affected @startergo and is also manifested here: https://sourceforge.net/p/refind/discussion/general/thread/8949634a88/

The manifestation varies but is there.

**Comment by dakanji** [comment]
X201 has the same tweak applied (more or less). Let's see if it make a difference
[X201-BOOTx64.efi.zip](https://github.com/dakanji/RefindPlus/files/8037278/X201-BOOTx64.efi.zip)

**Comment by startergo** [comment]
No boot.
[22d10b3622.log.zip](https://github.com/dakanji/RefindPlus/files/8037366/22d10b3622.log.zip)

**Comment by MarioG-X** [comment]
X201: Some progress:
two Opensuse on HD BTRFS boots fine
ElementaryOS on HD ext4 boots fine
Catalina on HD does not boot - screen does not change
Monterey on SDD does not boot
Catalina on SDD does not boot

So Linux boots but none of the macOSs boot.
[RFPLogs.zip](https://github.com/dakanji/RefindPlus/files/8037368/RFPLogs.zip)

**Comment by MarioG-X** [comment]
FYI don't know when it was changed but the firmware is 432.60.3.0.0 which shows as the latest for Minis with 12.2 or Catalina security 2022-01. So it must have been updated.

**Comment by dakanji** [comment]
Seems almost certain to be related to the firmware thing but possible there is an additional issue in play.
BTW, @startergo, I believe you can attach *.log files directly here.

Please try X201a to give a better baseline. Will pick up from there later
[X201a-BOOTx64.efi.zip](https://github.com/dakanji/RefindPlus/files/8037412/X201a-BOOTx64.efi.zip)

Thanks

**Comment by MarioG-X** [comment]
X201a: Still boots Linux, not macOS
[22d10b5505.log](https://github.com/dakanji/RefindPlus/files/8037446/22d10b5505.log)
[22d10b5614.log](https://github.com/dakanji/RefindPlus/files/8037448/22d10b5614.log)

**Comment by dakanji** [comment]
Could be that not freeing the memory addressed the firmware issue and that the MacOS failure is some other RP specific issue.

X199aV2 is tweaked to try to confirm this:
[X199aV2-BOOTx64.efi.zip](https://github.com/dakanji/RefindPlus/files/8037511/X199aV2-BOOTx64.efi.zip)

**Comment by MarioG-X** [comment]
X199aV2 is opposite of X201a, progress!:
Monterey on SSD boots!
Catalina on HD boots
Linux no longer boots
[22d10c2026.log](https://github.com/dakanji/RefindPlus/files/8037569/22d10c2026.log)
[22d10c2141.log](https://github.com/dakanji/RefindPlus/files/8037570/22d10c2141.log)
[22d10c2432.log](https://github.com/dakanji/RefindPlus/files/8037571/22d10c2432.log)
[22d10c2537.log](https://github.com/dakanji/RefindPlus/files/8037572/22d10c2537.log)

**Comment by dakanji** [comment]
Okay. Food for thought.
Will sleep on it and pick it up later.

Thanks for the help!

**Comment by MarioG-X** [comment]
X199aV3 the same as X199aV2:
[22d10c4655.log](https://github.com/dakanji/RefindPlus/files/8037650/22d10c4655.log)
[22d10c4744.log](https://github.com/dakanji/RefindPlus/files/8037651/22d10c4744.log)

**Comment by dakanji** [comment]
Thought about it a bit and decided it might be best to tackle things from the recent code and work backwards as opposed to trying to find the last working commit directly.

Please try X201aV2. No change from before apart from adding tracers to try to see where MacOS is crashing.
Took a look at the CVEs associated with recent updates and seems to be mainly "improved memory handling" and "improved boundary checking" and "fixing use-after-free issues".

Either Apple has gone overzealous or perhaps there is out-of-bounds stuff in the code. Coverity does flag instances which need to be looked into in detail. Will probably have to start upstream.

Anyway, here is X201aV2. Just additional logging as said.
[X201aV2-BOOTx64.efi.zip](https://github.com/dakanji/RefindPlus/files/8039614/X201aV2-BOOTx64.efi.zip)

**Comment by dakanji** [comment]
Sorry, I forgot to enable the additional logging for Level 0 in X201aV3
Updated the post with X201aV4 which has this if you don't mind.

No need to do X202

Thanks

**Comment by startergo** [comment]
Here you have it.
[22d10p5332.log](https://github.com/dakanji/RefindPlus/files/8041497/22d10p5332.log)

**Comment by dakanji** [comment]
X203 takes a stab at dealing with the hang point.
Hopefully it doesn't just go past this and hang further down the line
[X203-BOOTx64.efi.zip](https://github.com/dakanji/RefindPlus/files/8041884/X203-BOOTx64.efi.zip)

**Comment by dakanji** [comment]
Excellent.

Now, there was undesirable stuff, in the form of allowing memory leaks, needed and it would be best to try to minimise this.
So please try X204a & X204b if you don't mind and let me know which boots and which doesn't. Will do a few iterations of two files each until it starts failing on both. Alternatively, we might end up finding the actual critical line.
[X204_Archive.zip](https://github.com/dakanji/RefindPlus/files/8042493/X204_Archive.zip)

BTW, this is the REL file as we don't need logs. Just to know which boots and which doesn't.
Thanks

**Comment by dakanji** [comment]
Ok.  I suppose we best also make sure the REL version works in the first place before the split

If it does, might need to split into 3 instead of 2 before closing out.
Also, only need to test Monterey.

**Comment by dakanji** [comment]
X203a (REL Build of X203) attached
[X203a-BOOTx64.efi.zip](https://github.com/dakanji/RefindPlus/files/8042803/X203a-BOOTx64.efi.zip)

> Also, only need to test Monterey.

Probably better to test this fully as the DBG was

Thanks again!

**Comment by dakanji** [comment]
That's a bummer!

Well, it's back to the drawing board. Thought the workaround would be good until the actual issue could be found but looks like this has to be done now.

**Comment by dakanji** [comment]
Let's press on with trying to optimise the DBG build. Two files as with X204.
[X205_Archive.zip](https://github.com/dakanji/RefindPlus/files/8042923/X205_Archive.zip)

No need for logs and can just test Monterey.

**Comment by dakanji** [comment]
Hang on .... Very sorry.

The errors were mine.

**Comment by MarioG-X** [comment]
Both: Displays boot menu, Monterey selected, Press Enter, nothing happens, freeze, this time no white screen. Last lines in log:

  11:940    0:734  Processing Keystroke: UnicodeChar = 0x0D ... ScanCode = 0x00 - INFER_ENTER    Key:CarriageReturn
  11:943    0:003  Returned '1' (ENTER) from RunGenericMenu Call on 'Boot MacOS from FS Monterey' in 'RunMainMenu'
-------------------[ Load MacOS Instance ]-------------------

**Comment by dakanji** [comment]
OK. I switched git branches and didn't bring the fix along and that is why things were failing. Sorry about that.

Attaching X203aV2 to retest the REL build on the same basis as X203.
[X203aV2-BOOTx64.efi.zip](https://github.com/dakanji/RefindPlus/files/8043016/X203aV2-BOOTx64.efi.zip)

BTW, it might take quite a few iterations to nail things down so no rush and you can just pick thing up when you have a min. When I see a response from you, I'll move to the next step.

Thanks!

**Comment by dakanji** [comment]
Sorry to hear that ... Hope the trudge up the stairs is not too much strain on the back!

Attached revised X204. REL builds, Monterey only will do
[X204V2_Archive.zip](https://github.com/dakanji/RefindPlus/files/8043167/X204V2_Archive.zip)

**Comment by dakanji** [comment]
Must have mixed the files up as it was supposed to be a REL build.
Anyway, please try X205V2 ... just one file this time
[X205V2-BOOTx64.efi.zip](https://github.com/dakanji/RefindPlus/files/8043306/X205V2-BOOTx64.efi.zip)

**Comment by MarioG-X** [comment]
X205V2 boots Monterey fine. Did not try any other system.

**Comment by dakanji** [comment]
> Did not try any other system.

That's fine ... next one
[X206-BOOTx64.efi.zip](https://github.com/dakanji/RefindPlus/files/8043382/X206-BOOTx64.efi.zip)

**Comment by dakanji** [comment]
Nice ... Next one
[X207-BOOTx64.efi.zip](https://github.com/dakanji/RefindPlus/files/8043464/X207-BOOTx64.efi.zip)

**Comment by MarioG-X** [comment]
X207 boots Monterey fine. Did not try any other system.

**Comment by dakanji** [comment]
Lovely ... Next one
[X208-BOOTx64.efi.zip](https://github.com/dakanji/RefindPlus/files/8043623/X208-BOOTx64.efi.zip)

**Comment by MarioG-X** [comment]
X208 boots Monterey fine. Did not try any other system.

**Comment by dakanji** [comment]
Thanks ... X209 has three files
[X209_Archive.zip](https://github.com/dakanji/RefindPlus/files/8043705/X209_Archive.zip)

**Comment by MarioG-X** [comment]
X209 A, B and C boots Monterey fine. Did not try other systems.

**Comment by dakanji** [comment]
Hmmm .... that's odd. At least one of the three should have failed.
Can you reconfirm you didn't perhaps test the same file more than once?

**Comment by MarioG-X** [comment]
Each time I copied I was careful and checked the file size against the downloaded files. It was correct.

**Comment by dakanji** [comment]
Let's see what happens with X210.
[X210_Archive.zip](https://github.com/dakanji/RefindPlus/files/8043882/X210_Archive.zip)

This has both REL and DBG. Can you see what happens with both Linux and MacOS on both?

Thanks

**Comment by dakanji** [comment]
@startergo ... can you also test the two X210 files?

Will like to know whether one or boot or fail.

**Comment by dakanji** [comment]
Bizarre.

I switched off the whole allowing memory leaks thing that was needed to start with, since the three X209 files passing suggested it was not a factor. 

There was another change but that was only relevant to booting MacOS. Let's see what startergo finds.

Only possible explanation is that the multiple restarts you have done triggered NVRAM garbage collection.
Makes me wonder whether the MacOS change was even needed and all that was needed was a deep NVRAM reset.

Let me do a build with this change reverted

**Comment by dakanji** [comment]
Please try X211
[X211_Archive.zip](https://github.com/dakanji/RefindPlus/files/8044010/X211_Archive.zip)

**Comment by MarioG-X** [comment]
This time there are differences:
X211 REL boots linux fine, Monterey gets the white screen of death after pressing Enter on Monterey.
X211 DBG boots linux AND Monterey fine. Logs:
[22d10w2255.log](https://github.com/dakanji/RefindPlus/files/8044049/22d10w2255.log)
[22d10w2400.log](https://github.com/dakanji/RefindPlus/files/8044050/22d10w2400.log)

**Comment by MarioG-X** [comment]
I assure you each test was correct, for each test I made sure I used the correct version:
1. I moved the previous downloaded module(s) off the desktop.
2. In the EFI partition: I copied the name refind_x64.efi
3. In EFI I moved refind_x64.efi to the trash, most times emptied the trash.
4. I copied the latest downloaded module to the EFI leaving it's name so I could verify.
5. I verified I got the right name then renamed it to refind_x64.efi, I even checked file size to make sure they match.
6. Then tested.

Also, since this is elusive and you wondered if timing/NVRAM changes made a difference:
I retested X211 REL and DBG on all three linux (btrfs and ext4) and multiple macOS systems and all gave the same results as I reported for X211 DBG everything boots, X211 REL Linux ok, all macOS gets white screen freeze.

**Comment by MarioG-X** [comment]
I just reran the X209 A B C since you expected one to fail. I wondered if I booted Linux between each would it make a difference, it does not. The results are the same X209 A, B and C boot Linux and Monterey fine.

**Comment by MarioG-X** [comment]
I just reran X210 REL and DBG and got results as before. Both boot Linux and Monterey fine.

Since I retested all X209, X210 and X211 and got the same results as previously reported, this should eliminate booting sequence and installation as possible reasons for not getting the expected results. Sorry this is such a bugger but I will do what I can.

One of the worse bugs I had to find was: Spooler code goes to page dataset (swapfile), something modifies ONE BYTE on the swap file, later when needed that virtual page is swapped back in, BOOM! instruction fails and brings down entire mainframe! I found it but it only after loosing hair and having to use imagination. First thinking something modified memory, everyone assumed that. But no, a program modified the swap file. Obviously the problem was intermitted because it only showed up when that code was needed back in memory. Still gives me nightmares.

**Comment by startergo** [comment]
Both 210 boot and both 211 don't.

**Comment by dakanji** [comment]
Thanks both. Will take it forward on X210.
Still need to find and fix the root cause.

**Comment by dakanji** [comment]
> Still need to find and fix the root cause of how/why the firmware update is resulting in issues

A probable cause is that it doesn't like memory reallocation on some system table items.
Please try X212
[X212_Archive.zip](https://github.com/dakanji/RefindPlus/files/8047419/X212_Archive.zip)

**Comment by startergo** [comment]
This is no good.
[22d11r3610.log](https://github.com/dakanji/RefindPlus/files/8049796/22d11r3610.log)

**Comment by dakanji** [comment]
OK. Thanks both.

**Comment by dakanji** [comment]
Can you help with X213?
[X213-BOOTx64.efi.zip](https://github.com/dakanji/RefindPlus/files/8050677/X213-BOOTx64.efi.zip)


@startergo ... Can you run UEFI WIndows on this?

**Comment by MarioG-X** [comment]
X213 booted Opensuse fine and Montery fine. Those only 2 volumes tested.
[22d11t5359.log](https://github.com/dakanji/RefindPlus/files/8050741/22d11t5359.log)
[22d11t5757.log](https://github.com/dakanji/RefindPlus/files/8050742/22d11t5757.log)

**Comment by startergo** [comment]
Boots fine Windows and Monterey.
[22d11u1551.log](https://github.com/dakanji/RefindPlus/files/8050859/22d11u1551.log)
[22d11u1702.log](https://github.com/dakanji/RefindPlus/files/8050860/22d11u1702.log)

**Comment by dakanji** [comment]
OK. Wrapping up with that.
Will clean some stuff up and update later.
Thanks again!

**Comment by MarioG-X** [comment]
Standing by for anything you need to test. No problem.

**Comment by dakanji** [comment]
What seems to have happened is:
1. FAILURE TO BOOT MAC OS
RP has a function to which the BootServices->OpenProtocol is mapped. This ensures GOP is attached to the relevant interface when booting MacOS. This in turn ensures the Stage 1 loading screen is visible and is required by MacPros in some instances. I noticed something in the CVE notices for the recent updates about changes to prevent arbitrary code execution related to graphics drivers and presume this now halts loading when this remapping is done. I have now filtered the call to the function to exclude what I call "All-In-One" units with Apple Firmware:
    - Mac Books, iMacs and MacMinis
    - These units seem to always only have `Discounted [Other Item]` in the L I N K&nbsp;&nbsp;&nbsp;&nbsp;D E V I C E&nbsp;&nbsp;&nbsp;&nbsp;H A N D L E S section of the logs (unlike MacPros or other such without Apple Firmware).
    - Do not want to switch this functionality off on MacPros to accommodate these "All-In-One" units, so I hope it holds as I can't think of any other way to distinguish between them.
2. FAILURE TO BOOT LINUX
The ProtectNVRAM feature, which also involves similar remapping, was amended in an attempt to also catch Windows Installers/Updaters. It turned out that this was including Linux instances that do not use Grub. Not an issue ordinarily but I suppose the remapping touched on something the firmware guards against. These have now been filtered out.

I had actually, separately, filtered out these loaders Post Release of 0.13.2.AN but earlier in the process of testing for the issues, I had needed to switch the memory leaks on.

When I later switched to looking at things from the recent code, this filter was included and the leaks were no longer needed. We then went through the ultimately futile rounds of trying to narrow down a specific leak.

**Comment by dakanji** [comment]
@startergo ... One thing struck my mind. Can you confirm whether `spoof_osx_version` is a factor with UEFI Windows on X213? I see that you had this active.

**Comment by dakanji** [comment]
> No problem with the spoof_osx_version

Thanks but you still had it on as you did in the first test.

> I see that you had this active.

What I meant was to check whether it was a factor in your successful run by removing it.

**Comment by startergo** [comment]
That parameter has nothing to do with booting. It forces Intel on dual graphics laptops.

**Comment by startergo** [comment]
[22d13f3200.log](https://github.com/dakanji/RefindPlus/files/8055014/22d13f3200.log)
[22d13f3224.log](https://github.com/dakanji/RefindPlus/files/8055015/22d13f3224.log)
Boots fine with ```spoof_osx_version``` set to FALSE.

**Comment by MarioG-X** [comment]
FYI, I just tested on other iMac 27 2012 with Opensuse BTRFS SSD, Elementary OS EXT4 HD, and 2 macOS Catalinas HD and SSD. All good!
